### ISSUE TITLE
io.netty.resolver.dns.DnsNameResolver#doResolveCached race condition

### ISSUE THREAD

**Issue by johnou** [netty/netty#7583]
### Expected behavior

No exception.

### Actual behavior

```
2018-01-15 05:33:32,915 ERROR [Redisson-Epoll-1-15] (org.redisson.connection.DNSMonitor) Unable to resolve xxx.0001.use1.cache.amazonaws.com
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
    at java.util.ArrayList.rangeCheck(ArrayList.java:653) ~[?:1.8.0_144]
    at java.util.ArrayList.get(ArrayList.java:429) ~[?:1.8.0_144]
    at io.netty.resolver.dns.DnsNameResolver.doResolveCached(DnsNameResolver.java:613) ~[netty-resolver-dns-4.1.19.Final.jar:4.1.19.Final]
    at io.netty.resolver.dns.DnsNameResolver.doResolve(DnsNameResolver.java:593) ~[netty-resolver-dns-4.1.19.Final.jar:4.1.19.Final]
    at io.netty.resolver.dns.DnsNameResolver.doResolve(DnsNameResolver.java:527) ~[netty-resolver-dns-4.1.19.Final.jar:4.1.19.Final]
    at io.netty.resolver.SimpleNameResolver.resolve(SimpleNameResolver.java:63) ~[netty-resolver-4.1.19.Final.jar:4.1.19.Final]
    at io.netty.resolver.dns.InflightNameResolver.resolve(InflightNameResolver.java:100) ~[netty-resolver-dns-4.1.19.Final.jar:4.1.19.Final]
    at io.netty.resolver.dns.InflightNameResolver.resolve(InflightNameResolver.java:66) ~[netty-resolver-dns-4.1.19.Final.jar:4.1.19.Final]
    at io.netty.resolver.dns.InflightNameResolver.resolve(InflightNameResolver.java:51) ~[netty-resolver-dns-4.1.19.Final.jar:4.1.19.Final]
    at io.netty.resolver.InetSocketAddressResolver.doResolve(InetSocketAddressResolver.java:57) ~[netty-resolver-4.1.19.Final.jar:4.1.19.Final]
    at io.netty.resolver.InetSocketAddressResolver.doResolve(InetSocketAddressResolver.java:32) ~[netty-resolver-4.1.19.Final.jar:4.1.19.Final]
    at io.netty.resolver.AbstractAddressResolver.resolve(AbstractAddressResolver.java:108) ~[netty-resolver-4.1.19.Final.jar:4.1.19.Final]
    at org.redisson.connection.DNSMonitor$1.run(DNSMonitor.java:89) ~[redisson-3.5.7.jar:?]
    at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38) ~[netty-common-4.1.19.Final.jar:4.1.19.Final]
```

### Steps to reproduce

No reproducer.

### Minimal yet complete reproducer code (or URL to code)

n/a

### Netty version

4.1.19.Final

### JVM version (e.g. `java -version`)

java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

**Comment by johnou** [comment]
Some notes..

Might make more sense for io.netty.resolver.dns.DefaultDnsCache#get to return something like this to avoid concurrency issues with the list..

```java
    @Override
    public List<? extends DnsCacheEntry> get(String hostname, DnsRecord[] additionals) {
        checkNotNull(hostname, "hostname");
        if (!emptyAdditionals(additionals)) {
            return null;
        }
        return Collections.unmodifiableList(new ArrayList<DnsCacheEntry>(resolveCache.get(hostname)));
    }
```
otherwise we probably need to add a double check in the sync block to  io.netty.resolver.dns.DnsNameResolver#doResolveAllCached and io.netty.resolver.dns.DnsNameResolver#doResolveCached..

```java
    private boolean doResolveCached(String hostname,
                                    DnsRecord[] additionals,
                                    Promise<InetAddress> promise,
                                    DnsCache resolveCache) {
        final List<? extends DnsCacheEntry> cachedEntries = resolveCache.get(hostname, additionals);
        if (cachedEntries == null || cachedEntries.isEmpty()) {
            return false;
        }

        InetAddress address = null;
        Throwable cause = null;
        synchronized (cachedEntries) {
            final int numEntries = cachedEntries.size();
            if (numEntries == 0) {
                return false;
            }
...
```

**Comment by johnou** [comment]
cc @mrniko

**Comment by mrniko** [comment]
Seems it's a netty related bug? do you use 4.1.19 version?

**Comment by normanmaurer** [comment]
Fixed by https://github.com/netty/netty/pull/7584... return a new `ArrayList` will produce too much GC :(

**Comment by johnou** [comment]
@normanmaurer yeah figured as much, that's why I also mentioned the double check on the size in the sync block.

**Comment by johnou** [comment]
@normanmaurer thanks for the quick response!

**Comment by normanmaurer** [comment]
That said I think we may want to do something smarter in the long run. But I think the fix is ok for now.

**Comment by normanmaurer** [comment]
@johnou please check https://github.com/netty/netty/pull/7587 ... I think this is a more proper fix

**Comment by sindhuvaidy** [comment]
@normanmaurer : We are still experiencing this issue in netty 4.1.20 version.

**Comment by Scottmitch** [comment]
@sindhuvaidy - This is expected. The fix is included in 4.1.21. Please upgrade.
